### ISSUE TITLE
About  singletons

### ISSUE THREAD

**Issue by CriztianiX** [krakjoe/pthreads#465]
Please clarify my queries regarding Singleton and Multithreading:

What is the best way to implement Singleton  in a multithreaded environment?
What happens when multiple threads try to access getInstance() method at the same time?
Can we make singleton's getInstance() synchronized?
Is synchronization really needed, when using Singleton classes?

This is a piece of my code
```php
<?php
    class Task extends \Collectable {
        protected $page;
        private static $entity;

        public function __construct($page, $i) {
            $this->page = $page;
            $paginated = $i->paginate($page, $maxPerPage = 100);

            foreach ($paginated as $row) {
                $document = new \Couchbase\Document($row);
                $res = \Couchbase\Client::upsert($document->getId(), $document->toString());
            }
        }

        public function run() 
        {
            $this->setGarbage();
        }
    }
```

and here is the singleton class for couchbase
```php
<?php
namespace Couchbase 
{
  class Client 
  {
    public static $connection;
    public static function __callstatic($method, $args)
    {
      if (self::$connection) {
        return call_user_func_array(array(self::$connection, $method), $args);
      } else {
        $cluster = new \CouchbaseCluster('http://127.0.0.1:8091');
        self::$connection = $cluster->openBucket('default');

        if (self::$connection){
          return call_user_func_array(array(self::$connection, $method), $args);
        }
      }
    }
  }
}
```

This it seem works, but i dont know if it is implementend in the "best way"

**Comment by bnikola** [comment]
The best way to implement a singleton in multithreaded environment is not to use it. It's also questionable whether your use of threads is really a good choice. For some reason a lot of people here seem to use threads for absolutely wrong purposes, from curl to "simultaneous" connection to whatever, making everything slower and more difficult to deal with.

**Comment by cottton** [comment]
im not long into php pthreads
but it should work if you make sure the "target" is protected and extends Threaded.
What i see is that the "target" is somewhere in `call_user_func_array(array(self::$connection, $method), $args);`.
certainly `self::$connection` is an object.
so i could imagine:
```php
class {your connection} extends Threaded
{
    public function {your method}($args...)
    {
        // that method could be called by any context the same time
        // $this->arg = $args[...] so cut the work from here and call the protected below
        $this->onlyOneAtATime($args...);
    }
    protected function onlyOneAtATime($args...)
    {
        // as far i understood this now "breaks it down" 
        // and only one context can access that ethod at a time
        $this->arg = $args[...]
    }
}
```

**Comment by CriztianiX** [comment]
Good example!. So i should modifiy my actual implemantation

**Comment by bnikola** [comment]
The connection to an external API (Couchbase) is a **resource**. Resources cannot be serialized, therefore they cannot be shared. When two processes (or threads) try to access the same memory at the same time - you get a segmentation fault. This implies that you need to implement locking via mutual exclusion, if the sharing would even be possible. It's bad to have X threads access the same connection. You don't multiplex that way. Since you are after sharing a resource, and you're leaving the actual problem you're trying to solve - that's an XY problem. Your *perceived* solution Y is to share connection between threads, to solve your problem X. It won't work and it makes no sense either. You won't get speed, unless you consider segfaulting as fast program shutdown. Instead - rethink your problem and solve it without pthreads. What issue does parallelism solve for you in your use scenario?

**Comment by CriztianiX** [comment]
Ok. My scenario;
I need syncronize a postgresql database in couchbase cluster.
So first i run a count query to get the total rows to process, then
 i run a parallel query to get the paginated results and send them to couchbase.

Each thread run something like this:
 "SELECT * FROM abc LIMIT X, OFFSET X"

If this work is not in parallel, the syncronization can take hours depending of quantity the rows to process

**Comment by bnikola** [comment]
In the main context perform a count. After obtaining the count, determine how many threads you need.  Start the threads. Each thread **must** connect to Postgres in its `run` method. Other methods are insecure, might fail, might segfault. Doing it in `run` will work, 100%. Pass the limit / offset that each thread is supposed to handle. Have each thread fetch its part of data and send it to Couchbase. Once more, sending to Couchbase means that each thread is supposed to connect to it on its own. You can't share a resource such as database connection between threads using a singleton.

**Comment by CriztianiX** [comment]
Follow you recomendations, i've modified my code and this no loger works..
The paginated query no  get  any results, but any error is reported.
The strange things is moving the code in run() to __constructor() it works...

The complete code is:

```php
<?php
    class Task extends \Collectable {
        protected $page;

        public function __construct($page)
        {
            $this->page = $page;
        }

        public function run() 
        {

            $paginated = \MundoDescuento\ApplicationUsersQuery::create()->paginate($this->page, $maxPerPage = 100);

            var_dump($paginated); // no results, or error.

            foreach ($paginated as $row) 
            {
                $cluster = new \CouchbaseCluster('http://127.0.0.1:8091');
                $bucket = $cluster->openBucket('default');
                $document = new \Couchbase\Document($row);
                $res = $bucket->upsert($document->getId(), $document->toString());
            }

            $this->setGarbage();
        }
    }
```

**Comment by bnikola** [comment]
The error is probably due to the fact that thread in its run method hasn't got the access to your autoloader so it can't load the classes you invoked. Constructor does have the access to those, but once run is invoked - that's the point where you execute data in the tread context. Long story short: include your autoloader in your `run` method.

**Comment by CriztianiX** [comment]
Ok. I've re-make all job again.

In this respository you can see it:
https://github.com/CriztianiX/rds-couchbase-syncronizer

This seems works, A litle slow performance, (over 2k operations per seconds) but...it works.
Any suggest?

PD: Still needed a code refactoring!

**Comment by cottton** [comment]
Not sure but maybe you can reach more if you increase the `protected static $limit = 100;`.
The Worker/Collectable looks good ima.
But ofc you now open for every Task (Collectable) a new connection (not wrong) for only for 100 rows (when im not mistaken).
Means - new Task, new conn, 100 rows ... oh, already done ... =)
 
The work on the given rows (limit) should at least take as much time it takes to open the connection.

try that:
```php
<?php
namespace Rds {
    class Task extends \Collectable {
        protected $offset;
        protected $limit;
        public function __construct($offset, $limit) {
            $this->offset = $offset;
            $this->limit = $limit;
        }
        public function run()
        {
          $a = microtime(true);

          /* Get new spot2 conection */
          $mapper = \Rds\Mapper::getMapper();
          $rows = $mapper->query("SELECT * FROM city WHERE id > :id LIMIT :limit OFFSET :offset", [
            'id' => 0,
            'limit' => $this->limit,
            'offset' => $this->offset
          ]);
          /* Get new conection for couchbase bucket */
          $bucket = \Rds\CouchbaseConnection::getBucketConnection();

          $a = microtime(true)-$a;

          $b = microtime(true);

          foreach ($rows as $row) {
            $tmp = [];
            $tmp["id"] = $row->id;
            $tmp["name"] = $row->name;
            $res = $bucket->upsert((string)$row->id, json_encode($tmp));
          }

          $b = microtime(true)-$a;

          echo round($this->limit * $a / $b); /* set protected static $limit to that val 
                                              to spend as much time for rows 
                                              as you spend to create conn ect 
                                              ... or * 2 ? */

          $this->setGarbage();
        }
    }
}
```
easy example:
```php
$limit = 1235;

$a = microtime(true);
usleep(1234567); // simulate conn ect
$a = microtime(true)-$a;

$b = microtime(true);
for($i = 0; $i < $limit; $i++){
    usleep(123); // work
}
$b = microtime(true)-$b;

echo 'time spent conn: ' . $a . PHP_EOL;
echo 'time spent work: ' . $b . PHP_EOL;
echo 'set limit to ' . round($limit * $a / $b) . ' to get it even';

/* 1st run:
time spent conn: 1.2350699901581
time spent work: 0.10000610351562    <--- does not justify the time we spent on conn
set limit to 1235 to get it even

set limit to 1235:
time spent conn: 1.2350709438324
time spent work: 1.2600719928741
set limit to 1210 to get it even
*/
```

**Comment by CriztianiX** [comment]
Seeing the performance something is not clear .
I think I'm having "time- slicing".
The green line is "optimus performace" when i execute the same  program written in JAVA.
Any idea?
![couchbase php](https://cloud.githubusercontent.com/assets/1068190/9071719/3098c83e-3acd-11e5-838d-264a225cdcbb.png)

**Comment by cottton** [comment]
I tihnk thats normal. You got 8 workers running and all of them got similar work.
They all finish their Tasks more or less the same time and start working on the next Task.

You would get the "perfect speed" if you set as much worker (new Pool(n, ,,,)) as you got Tasks.
Then they all would work on 1 Task at a time. But ofc this wont be possible most time.
Just try to set as many Workers as you can,

Maybe you can decrease the "downtime" (the time each Task takes to create the 2 connections) if you create the 2 connections you need at the Workers.
See: https://github.com/krakjoe/pthreads/blob/master/examples/SQLWorker.php
That way each Worker would create its own reusable connections.
The Task then will be able to use `$this->worker->mapper->query(...)` and `$this->worker->bucket->upsert(...)`.

Also - before you change anything - are you able to use persistant connections?

**Comment by CriztianiX** [comment]
> You would get the "perfect speed" if you set as much worker (new Pool(n, ,,,)) as you got Tasks.
> Then they all would work on 1 Task at a time. But ofc this wont be possible most time.
> Just try to set as many Workers as you can,
are there  a way to calcute the workers count  for each system?

**Comment by CriztianiX** [comment]
anyone have any suggestions because I get this error only using pthreads?

**Comment by krakjoe** [comment]
This is actually not the place to get help with programming; I can't see the issue you are reporting.

If there is an actual issue, please open a new report and be specific.
